core/src/main/scala/quasar/destination/s3/impl/DefaultUpload.scala

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2014–2019 SlamData Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.destination.s3.impl
+
+import quasar.destination.s3.{Bucket, ObjectKey, Upload}
+
+import slamdata.Predef._
+
+import cats.effect.{Concurrent, ExitCase}
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import fs2.Stream
+import monix.catnap.syntax._
+import software.amazon.awssdk.services.s3.S3AsyncClient
+import software.amazon.awssdk.core.async.AsyncRequestBody
+import software.amazon.awssdk.services.s3.S3AsyncClient
+import software.amazon.awssdk.services.s3.model.{
+  AbortMultipartUploadRequest,
+  AbortMultipartUploadResponse,
+  CompleteMultipartUploadRequest,
+  CompleteMultipartUploadResponse,
+  CompletedMultipartUpload,
+  CompletedPart,
+  CreateMultipartUploadRequest,
+  CreateMultipartUploadResponse,
+  UploadPartRequest
+}
+
+final case class DefaultUpload[F[_]: Concurrent](client: S3AsyncClient, partSize: Int) extends Upload[F] {
+  def upload(bytes: Stream[F, Byte], bucket: Bucket, key: ObjectKey): F[Unit] =
+    Concurrent[F].bracketCase(
+      startUpload(client, bucket, key))(createResponse =>
+      for {
+        parts <- uploadParts(client, bytes, createResponse.uploadId, partSize, bucket, key)
+        _ <- completeUpload(client, createResponse.uploadId, bucket, key, parts)
+      } yield ()) {
+      case (createResponse, ExitCase.Canceled | ExitCase.Error(_)) =>
+        abortUpload(client, createResponse.uploadId, bucket, key).void
+      case (_, ExitCase.Completed) =>
+        Concurrent[F].unit
+    }
+
+  private def startUpload(client: S3AsyncClient, bucket: Bucket, key: ObjectKey)
+      : F[CreateMultipartUploadResponse] =
+    Concurrent[F].delay(client.createMultipartUpload(
+      CreateMultipartUploadRequest.builder.bucket(bucket.value).key(key.value).build)).futureLift
+
+  private def uploadParts(
+    client: S3AsyncClient,
+    bytes: Stream[F, Byte],
+    uploadId: String,
+    minChunkSize: Int,
+    bucket: Bucket,
+    key: ObjectKey): F[List[CompletedPart]] =
+    (bytes.chunkMin(minChunkSize).zipWithIndex evalMap {
+      case (byteChunk, n) => {
+        // parts numbers must start at 1
+        val partNumber = Int.box(n.toInt + 1)
```

Comment:
Why the explicit boxing?

Reply:
Not sure. Implicit conversions are triggered when I import `scala.Predef._`, but don't work when I use `slamdata.Predef._`.

Reply:
Hmm. Seems like `int2Integer` is available in `scala.Predef` but not in `slamdata.Predef`